<!doctype html>
<html class="no-js" lang="en" data-content_root="../../">
  <head><meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Soft Segmentation Tutorial (Full version)" />
<meta property="og:type" content="website" />
<meta property="og:url" content="tutorials/notebooks/Tutorial_Mousebrian_full.html" />
<meta property="og:site_name" content="SMURF" />
<meta property="og:description" content="Here, we provide the full version of cell segmentation for SMURF using the FFPE Mouse Brain data from 10x. Notably, this version requires GPU resources and please install the full version of SMURF...." />
<meta property="og:image:width" content="1146" />
<meta property="og:image:height" content="600" />
<meta property="og:image" content="_images/social_previews/summary_tutorials_notebooks_Tutorial_Mousebrian_full_3c8c1578.png" />
<meta property="og:image:alt" content="Here, we provide the full version of cell segmentation for SMURF using the FFPE Mouse Brain data from 10x. Notably, this version requires GPU resources and please install the full version of SMURF...." />
<meta name="description" content="Here, we provide the full version of cell segmentation for SMURF using the FFPE Mouse Brain data from 10x. Notably, this version requires GPU resources and please install the full version of SMURF...." />
<meta name="twitter:card" content="summary_large_image" />
<link rel="index" title="Index" href="../../genindex.html"><link rel="search" title="Search" href="../../search.html"><link rel="next" title="Unrolling Tutorial" href="unroll_example.html"><link rel="prev" title="Soft Segmentation Tutorial (Lite version)" href="Tutorial_Mousebrian.html">

    <!-- Generated with Sphinx 8.1.3 and Furo 2025.09.25 -->
        <title>Soft Segmentation Tutorial (Full version) - SMURF</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=580074bf" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.custom.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster.bundle.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-shadow.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-punk.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-noir.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-light.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/tooltipster-sideTip-borderless.min.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/css/micromodal.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=8dab3a3b" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  --color-brand-primary: #003262;
  --color-brand-content: #003262;
  --admonition-font-size: var(--font-size-normal);
  --admonition-title-font-size: var(--font-size-normal);
  --code-font-size: var(--font-size--small);
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle site navigation sidebar">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc" aria-label="Toggle table of contents sidebar">
<label class="overlay sidebar-overlay" for="__navigation"></label>
<label class="overlay toc-overlay" for="__toc"></label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <span class="icon"><svg><use href="#svg-menu"></use></svg></span>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">SMURF</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  <span class="sidebar-brand-text">SMURF</span>
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../index.html">Tutorials</a><input aria-label="Toggle navigation of Tutorials" checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Tutorial_Mousebrian.html">Soft Segmentation Tutorial (Lite version)</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Soft Segmentation Tutorial (Full version)</a></li>
<li class="toctree-l2"><a class="reference internal" href="unroll_example.html">Unrolling Tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="unroll_example.html#Unrolling-Mouse-Brain-Data">Unrolling Mouse Brain Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="unroll_example.html#Unrolling-Mouse-Small-Intestine-Data">Unrolling Mouse Small Intestine Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="Tutorial_cell_segmentation.html">Nuclei Segmentation by StarDist</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../Data_structure/index.html">Data Structure</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/index.html">API</a><input aria-label="Toggle navigation of API" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/smurf.html">SMURF</a><input aria-label="Toggle navigation of SMURF" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.calculate_weight_to_celltype.html">smurf.calculate_weight_to_celltype</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.expanding_cells.html">smurf.expanding_cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.get_finaldata.html">smurf.get_finaldata</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.get_finaldata_fast.html">smurf.get_finaldata_fast</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.itering_arragement.html">smurf.itering_arragement</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.make_pixels_cells.html">smurf.make_pixels_cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.make_preparation.html">smurf.make_preparation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.nuclei_rna.html">smurf.nuclei_rna</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.plot_cellcluster_position.html">smurf.plot_cellcluster_position</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.plot_results.html">smurf.plot_results</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.prepare_dataframe_image.html">smurf.prepare_dataframe_image</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.return_celltype_plot.html">smurf.return_celltype_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.singlecellanalysis.html">smurf.singlecellanalysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.start_optimization.html">smurf.start_optimization</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../api/spatial_object.html">Spatial Object</a><input aria-label="Toggle navigation of Spatial Object" class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><span class="icon"><svg><use href="#svg-arrow-right"></use></svg></span></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.spatial_object.spatial_object.html">smurf.spatial_object.spatial_object</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.spatial_object.spatial_object.add_segmentation.html">smurf.spatial_object.spatial_object.add_segmentation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.spatial_object.spatial_object.generate_cell_spots_information.html">smurf.spatial_object.spatial_object.generate_cell_spots_information</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../api/reference/smurf.spatial_object.spatial_object.image_temp.html">smurf.spatial_object.spatial_object.image_temp</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../release_notes/index.html">Release notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../references.html">References</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/The-Mitra-Lab/SMURF">GitHub</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="../../_sources/tutorials/notebooks/Tutorial_Mousebrian_full.ipynb.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle" aria-label="Toggle Light / Dark / Auto color theme">
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <span class="icon"><svg><use href="#svg-toc"></use></svg></span>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="Soft-Segmentation-Tutorial-(Full-version)">
<h1>Soft Segmentation Tutorial (Full version)<a class="headerlink" href="#Soft-Segmentation-Tutorial-(Full-version)" title="Link to this heading">¶</a></h1>
<p>Here, we provide the full version of cell segmentation for SMURF using the FFPE Mouse Brain data from 10x. Notably, this version requires GPU resources and please install the full version of SMURF.</p>
<section id="Dataset-Information">
<h2>Dataset Information<a class="headerlink" href="#Dataset-Information" title="Link to this heading">¶</a></h2>
<p>Access the dataset here: <a class="reference external" href="https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-brain-he">Visium HD CytAssist Gene Expression Libraries of Mouse Brain</a></p>
</section>
<section id="Important-Downloads">
<h2>Important Downloads<a class="headerlink" href="#Important-Downloads" title="Link to this heading">¶</a></h2>
<p>To download the necessary files, run the following commands:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Download data from 10x</span>
<span class="err">!</span><span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="s2">&quot;smurf_example_data&quot;</span>
<span class="err">!</span><span class="n">wget</span> <span class="o">-</span><span class="n">P</span> <span class="n">smurf_example_data</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">cf</span><span class="mf">.10</span><span class="n">xgenomics</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">samples</span><span class="o">/</span><span class="n">spatial</span><span class="o">-</span><span class="n">exp</span><span class="o">/</span><span class="mf">3.0.0</span><span class="o">/</span><span class="n">Visium_HD_Mouse_Brain</span><span class="o">/</span><span class="n">Visium_HD_Mouse_Brain_tissue_image</span><span class="o">.</span><span class="n">tif</span>
<span class="err">!</span><span class="n">wget</span> <span class="o">-</span><span class="n">P</span> <span class="n">smurf_example_data</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">cf</span><span class="mf">.10</span><span class="n">xgenomics</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">samples</span><span class="o">/</span><span class="n">spatial</span><span class="o">-</span><span class="n">exp</span><span class="o">/</span><span class="mf">3.0.0</span><span class="o">/</span><span class="n">Visium_HD_Mouse_Brain</span><span class="o">/</span><span class="n">Visium_HD_Mouse_Brain_binned_outputs</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
<span class="err">!</span><span class="n">tar</span> <span class="o">-</span><span class="n">xzvf</span> <span class="n">smurf_example_data</span><span class="o">/</span><span class="n">Visium_HD_Mouse_Brain_binned_outputs</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">C</span> <span class="n">smurf_example_data</span>

<span class="c1"># Download our nuclei segmentation results if needed</span>
<span class="err">!</span><span class="n">curl</span> <span class="o">-</span><span class="n">L</span> <span class="o">-</span><span class="n">o</span> <span class="n">smurf_example_data</span><span class="o">/</span><span class="n">segmentation_final</span><span class="o">.</span><span class="n">npy</span><span class="o">.</span><span class="n">gz</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">The</span><span class="o">-</span><span class="n">Mitra</span><span class="o">-</span><span class="n">Lab</span><span class="o">/</span><span class="n">SMURF</span><span class="o">/</span><span class="n">raw</span><span class="o">/</span><span class="n">main</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">segmentation_final</span><span class="o">.</span><span class="n">npy</span><span class="o">.</span><span class="n">gz</span>
<span class="err">!</span><span class="n">gunzip</span> <span class="n">smurf_example_data</span><span class="o">/</span><span class="n">segmentation_final</span><span class="o">.</span><span class="n">npy</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>If you use the same cell segmentation method as introduced, you may skip the first part until the section labeled <a class="reference external" href="https://the-mitra-lab.github.io/SMURF/tutorials/notebooks/Tutorial_Mousebrian_full.html#Start-Here">Start here</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">smurf</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">su</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">copy</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">gzip</span>
</pre></div>
</div>
</div>
<blockquote>
<div><p><strong>Note on software versions</strong></p>
<div class="line-block">
<div class="line">This tutorial was developed and tested with the exact package versions listed below.</div>
<div class="line">Using these versions should reproduce the results exactly.</div>
<div class="line">Newer releases may introduce subtle differences.</div>
<div class="line">If you encounter any issues, please open an issue on our GitHub repository and we’ll respond as soon as possible. Thanks!</div>
</div>
</div></blockquote>
<div class="table-wrapper docutils container">
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Package</p></th>
<th class="head"><p>Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>numpy</p></td>
<td><p>1.26.4</p></td>
</tr>
<tr class="row-odd"><td><p>pandas</p></td>
<td><p>1.5.3</p></td>
</tr>
<tr class="row-even"><td><p>tqdm</p></td>
<td><p>4.66.2</p></td>
</tr>
<tr class="row-odd"><td><p>scanpy</p></td>
<td><p>1.10.0</p></td>
</tr>
<tr class="row-even"><td><p>scipy</p></td>
<td><p>1.12.0</p></td>
</tr>
<tr class="row-odd"><td><p>matplotlib</p></td>
<td><p>3.8.3</p></td>
</tr>
<tr class="row-even"><td><p>numba</p></td>
<td><p>0.59.0</p></td>
</tr>
<tr class="row-odd"><td><p>scikit-learn</p></td>
<td><p>1.4.1.post1</p></td>
</tr>
<tr class="row-even"><td><p>Pillow</p></td>
<td><p>10.2.0</p></td>
</tr>
<tr class="row-odd"><td><p>anndata</p></td>
<td><p>0.10.6</p></td>
</tr>
<tr class="row-even"><td><p>h5py</p></td>
<td><p>3.10.0</p></td>
</tr>
<tr class="row-odd"><td><p>pyarrow</p></td>
<td><p>16.1.0</p></td>
</tr>
<tr class="row-even"><td><p>igraph</p></td>
<td><p>0.11.5</p></td>
</tr>
<tr class="row-odd"><td><p>torch</p></td>
<td><p>2.2.1+cu12</p></td>
</tr>
</tbody>
</table>
</div>
<p>Here, we define your data path and saving path. Ensure that your data path includes the complete HE or DAPI image and a folder containing <code class="docutils literal notranslate"><span class="pre">square_002um</span></code> data. Specifically, you will need <code class="docutils literal notranslate"><span class="pre">tissue_positions.parquet</span></code> and <code class="docutils literal notranslate"><span class="pre">filtered_feature_bc_matrix</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="n">data_path</span> <span class="o">=</span> <span class="s1">&#39;smurf_example_data/&#39;</span>
<span class="n">save_path</span> <span class="o">=</span> <span class="s1">&#39;results_example_mousebrain_full/&#39;</span>

<span class="c1"># Check if the directory exists</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">save_path</span><span class="p">):</span>
    <span class="c1"># If it doesn&#39;t exist, create it</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">save_path</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>We start building our So (spatial object). The input should be your <code class="docutils literal notranslate"><span class="pre">tissue_positions.parquet</span></code> in <code class="docutils literal notranslate"><span class="pre">'square_002um/spatial'</span></code>, the final full image, and the format of your experiment (either <code class="docutils literal notranslate"><span class="pre">'HE'</span></code> or <code class="docutils literal notranslate"><span class="pre">'DAPI'</span></code>).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">so</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">prepare_dataframe_image</span><span class="p">(</span><span class="n">data_path</span><span class="o">+</span><span class="s1">&#39;binned_outputs/square_002um/spatial/tissue_positions.parquet&#39;</span><span class="p">,</span>
                                <span class="n">data_path</span><span class="o">+</span><span class="s1">&#39;Visium_HD_Mouse_Brain_tissue_image.tif&#39;</span><span class="p">,</span>
                               <span class="s1">&#39;HE&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Please use your own tools to perform cell segmentation on <code class="docutils literal notranslate"><span class="pre">so.image_temp()</span></code> (the area covered by Visium HD in the original image). Set all background pixels to <code class="docutils literal notranslate"><span class="pre">0</span></code> and assign each cell a unique integer cell ID. <code class="docutils literal notranslate"><span class="pre">segmentation_final</span></code> should be a NumPy array with the same shape as <code class="docutils literal notranslate"><span class="pre">so.image_temp()</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">image_temp</span><span class="p">())</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_8_0.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_8_0.png" />
</div>
</div>
<p>Please save the image for segmentation now</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">GzipFile</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span> <span class="s1">&#39;image_to_segmentation.npy.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">so</span><span class="o">.</span><span class="n">image_temp</span><span class="p">())</span>
</pre></div>
</div>
</div>
<p>Assuming we now have the final segmentation results, let’s proceed to read them. You can always use the provided example results to try out the tutorial.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">so</span><span class="o">.</span><span class="n">segmentation_final</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">data_path</span><span class="o">+</span><span class="s1">&#39;segmentation_final.npy&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now, we start generating cell and spot information to prepare for the analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">so</span><span class="o">.</span><span class="n">generate_cell_spots_information</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating cells information.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 23340/23340 [01:06&lt;00:00, 352.34it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Generating spots information.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 23340/23340 [02:51&lt;00:00, 135.74it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Filtering cells
We have 57300 nuclei in total.
Creating NN network
</pre></div></div>
</div>
<p>Visualize the nuclei segmentation results overlaid on the original HE or DAPI image. Ensure that the results are meaningful and that the segmented nuclei accurately align with the original image.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">su</span><span class="o">.</span><span class="n">plot_results</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">image_temp</span><span class="p">(),</span> <span class="n">so</span><span class="o">.</span><span class="n">segmentation_final</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_16_0.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_16_0.png" />
</div>
</div>
<section id="Start-Here">
<h3>Start Here<a class="headerlink" href="#Start-Here" title="Link to this heading">¶</a></h3>
<p>if you use the same cell segmentation method as introduced:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">smurf</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">su</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">copy</span>


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;so.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">so</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
<p>Load <code class="docutils literal notranslate"><span class="pre">adata</span></code> from the original the 2um results.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># load adata from original 2um results</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_mtx</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span><span class="s1">&#39;binned_outputs/square_002um/filtered_feature_bc_matrix&#39;</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">so</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="n">so</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">in_tissue</span> <span class="o">==</span> <span class="mi">1</span><span class="p">][</span><span class="s1">&#39;barcode&#39;</span><span class="p">]])</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>s_002um_02448_01644-1</th>
    </tr>
    <tr>
      <th>s_002um_00700_02130-1</th>
    </tr>
    <tr>
      <th>s_002um_00305_01021-1</th>
    </tr>
    <tr>
      <th>s_002um_02703_00756-1</th>
    </tr>
    <tr>
      <th>s_002um_02278_00850-1</th>
    </tr>
    <tr>
      <th>...</th>
    </tr>
    <tr>
      <th>s_002um_02189_01783-1</th>
    </tr>
    <tr>
      <th>s_002um_01609_01578-1</th>
    </tr>
    <tr>
      <th>s_002um_00456_02202-1</th>
    </tr>
    <tr>
      <th>s_002um_00292_01011-1</th>
    </tr>
    <tr>
      <th>s_002um_00865_01560-1</th>
    </tr>
  </tbody>
</table>
<p>6296688 rows × 0 columns</p>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_counts</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here, we begin by generating the <code class="docutils literal notranslate"><span class="pre">nuclei*genes</span></code> matrix. The result will be stored in <code class="docutils literal notranslate"><span class="pre">so.final_nuclei</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">su</span><span class="o">.</span><span class="n">nuclei_rna</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">so</span><span class="p">)</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">final_nuclei</span><span class="p">)</span>
<span class="n">adata_sc</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 57300/57300 [01:08&lt;00:00, 836.90it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 57300 × 9970
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;
</pre></div></div>
</div>
<p>Filter cells by counts while preserving the raw data version.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">min_counts</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">adata_raw</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">)</span>
<span class="n">adata_raw</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 56867 × 9970
    obs: &#39;n_counts&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_raw</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Initial analysis of nuclei data. Note that many datasets may not perform well at this stage; it is normal to observe hundreds of clusters due to low-quality nuclei counts.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">singlecellanalysis</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span><span class="n">resolution</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/juanru/miniconda3/envs/bidcell/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting Clustering
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_27_3.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_27_3.png" />
</div>
</div>
<p>Here, we start performing iterations. Note that this function can be time-consuming.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">su</span><span class="o">.</span><span class="n">itering_arragement</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">adata_raw</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">so</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">save_folder</span> <span class="o">=</span> <span class="n">save_path</span><span class="p">,</span> <span class="n">show</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">keep_previous</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [03:48&lt;00:00, 249.32it/s]
100%|██████████| 56867/56867 [01:05&lt;00:00, 874.60it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
Starting Clustering
WARNING: saving figure to file figures/umap1_umap.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_2.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_2.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.42587605351650787 1
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [03:50&lt;00:00, 246.54it/s]
100%|██████████| 56867/56867 [01:02&lt;00:00, 904.00it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
Starting Clustering
WARNING: saving figure to file figures/umap2_umap.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_6.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_6.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.6653163073440685 2
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [04:40&lt;00:00, 202.72it/s]
100%|██████████| 56867/56867 [01:03&lt;00:00, 900.55it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
Starting Clustering
WARNING: saving figure to file figures/umap3_umap.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_10.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_10.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.692272176364526 3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [04:19&lt;00:00, 219.06it/s]
100%|██████████| 56867/56867 [01:02&lt;00:00, 903.81it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
Starting Clustering
WARNING: saving figure to file figures/umap4_umap.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_14.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_14.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.7058946412586965 4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [04:34&lt;00:00, 206.80it/s]
100%|██████████| 56867/56867 [01:02&lt;00:00, 903.52it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
Starting Clustering
WARNING: saving figure to file figures/umap5_umap.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_18.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_18.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.7017089650355296 5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [04:50&lt;00:00, 195.66it/s]
100%|██████████| 56867/56867 [01:03&lt;00:00, 897.56it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting mt
Starting normalization
Starting PCA
Starting UMAP
Starting Clustering
WARNING: saving figure to file figures/umap6_umap.png
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_22.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_29_22.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.6986723859819448 6
</pre></div></div>
</div>
<p>Load results here. Note that the full version differs from the simple version from this point onward. This is your last chance to switch to the lite version. (Note that <code class="docutils literal notranslate"><span class="pre">adatas_final</span></code>, <code class="docutils literal notranslate"><span class="pre">cells_final</span></code> and <code class="docutils literal notranslate"><span class="pre">weights_record</span></code> can be used directly there.)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adatas_final</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span><span class="s1">&#39;adatas.h5ad&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span><span class="s1">&#39;cells_final.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">cells_final</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span><span class="s1">&#39;weights_record.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">weights_record</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here, we generate the following:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pct_toml_dic</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spots_X_dic</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">celltypes_dic</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cells_X_plus_dic</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">nonzero_indices_dic</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">nonzero_indices_toml</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cells_before_ml</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cells_before_ml_x</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">groups_combined</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spots_id_dic</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spots_id_dic_prop</span></code></p></li>
</ul>
<p>These are the results used as input to the deep learning model to determine the final proportion of each spot. Adjust <code class="docutils literal notranslate"><span class="pre">maximum_cells</span> <span class="pre">=</span> <span class="pre">6000</span></code> based on your GPU memory capacity</p>
<p>If you have millions of cells, the process may take a considerable amount of time. You can save these files if needed.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1">#save files:</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">gzip</span>

<span class="k">def</span><span class="w"> </span><span class="nf">save_compressed_pickle</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">filename</span><span class="p">):</span>
    <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">GzipFile</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">HIGHEST_PROTOCOL</span><span class="p">)</span>


<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">spots_X_dic</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;spots_X_dic.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">celltypes_dic</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;celltypes_dic.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">nonzero_indices_dic</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;nonzero_indices_dic.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">nonzero_indices_toml</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;nonzero_indices_toml.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">cells_X_plus_dic</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;cells_X_plus_dic.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">cells_before_ml</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;cells_before_ml.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">cells_before_ml_x</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;cells_before_ml_x.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">groups_combined</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;groups_combined.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">spots_id_dic</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;spots_id_dic.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">spots_id_dic_prop</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;spots_id_dic_prop.pkl.gz&quot;</span><span class="p">)</span>
<span class="n">save_compressed_pickle</span><span class="p">(</span><span class="n">pct_toml_dic</span><span class="p">,</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;pct_toml_dic.pkl.gz&#39;</span><span class="p">)</span>


<span class="c1">#load files:</span>
<span class="c1"># Define a function to load compressed pickles</span>
<span class="k">def</span><span class="w"> </span><span class="nf">load_compressed_pickle</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">GzipFile</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

<span class="n">weight_to_celltype</span> <span class="o">=</span> <span class="n">load_compressed_pickle</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;weight_to_celltype.pkl.gz&#39;</span><span class="p">)</span>
<span class="n">cells_before_ml</span> <span class="o">=</span> <span class="n">load_compressed_pickle</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;cells_before_ml.pkl.gz&#39;</span><span class="p">)</span>
<span class="n">groups_combined</span> <span class="o">=</span> <span class="n">load_compressed_pickle</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;groups_combined.pkl.gz&#39;</span><span class="p">)</span>
<span class="n">pct_toml_dic</span> <span class="o">=</span> <span class="n">load_compressed_pickle</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;pct_toml_dic.pkl.gz&#39;</span><span class="p">)</span>
<span class="n">nonzero_indices_dic</span> <span class="o">=</span> <span class="n">load_compressed_pickle</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;nonzero_indices_dic.pkl.gz&#39;</span><span class="p">)</span>
<span class="n">nonzero_indices_toml</span> <span class="o">=</span> <span class="n">load_compressed_pickle</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;nonzero_indices_toml.pkl.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pct_toml_dic</span><span class="p">,</span> <span class="n">spots_X_dic</span><span class="p">,</span> <span class="n">celltypes_dic</span><span class="p">,</span> <span class="n">cells_X_plus_dic</span><span class="p">,</span> <span class="n">nonzero_indices_dic</span><span class="p">,</span> <span class="n">nonzero_indices_toml</span><span class="p">,</span> <span class="n">cells_before_ml</span><span class="p">,</span> <span class="n">cells_before_ml_x</span><span class="p">,</span> <span class="n">groups_combined</span><span class="p">,</span> <span class="n">spots_id_dic</span><span class="p">,</span><span class="n">spots_id_dic_prop</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">make_preparation</span><span class="p">(</span><span class="n">cells_final</span><span class="p">,</span>
                                                                                                                <span class="n">so</span><span class="p">,</span> <span class="n">adatas_final</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">weights_record</span><span class="p">,</span> <span class="n">maximum_cells</span> <span class="o">=</span> <span class="mi">6000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  4%|▍         | 69727/1690156 [00:05&lt;02:09, 12539.06it/s]/home/juanru/miniconda3/envs/bidcell/lib/python3.10/site-packages/scipy/sparse/_base.py:699: RuntimeWarning: divide by zero encountered in divide
  return self.astype(np.float64)._mul_scalar(1./other)
100%|██████████| 1690156/1690156 [01:51&lt;00:00, 15152.36it/s]
</pre></div></div>
</div>
<p>Start preparing CUDA for analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>
<span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda&quot;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>
<span class="n">device</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
device(type=&#39;cuda&#39;)
</pre></div></div>
</div>
<p>Start learning… Save results immediately! Typically, training one group takes about 10 minutes for NVIDIA GeForce RTX 3090.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">spot_cell_dic</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">start_optimization</span><span class="p">(</span><span class="n">spots_X_dic</span><span class="p">,</span> <span class="n">celltypes_dic</span><span class="p">,</span> <span class="n">cells_X_plus_dic</span><span class="p">,</span> <span class="n">nonzero_indices_toml</span><span class="p">,</span> <span class="n">device</span><span class="p">,</span>
                                      <span class="n">num_epochs</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">print_each</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">epsilon</span><span class="o">=</span><span class="mf">0.00001</span><span class="p">,)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span><span class="s1">&#39;spot_cell_dic.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">spot_cell_dic</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Group 0:
Epoch 1, Loss: 0.3875774145126343
Stopping early at epoch 31 due to minimal loss improvement.
Group 1:
Epoch 1, Loss: 0.5505884885787964
Stopping early at epoch 35 due to minimal loss improvement.
Group 2:
Epoch 1, Loss: 0.5124831199645996
Stopping early at epoch 37 due to minimal loss improvement.
Group 3:
Epoch 1, Loss: 0.45128583908081055
Stopping early at epoch 32 due to minimal loss improvement.
Group 4:
Epoch 1, Loss: 0.38966405391693115
Stopping early at epoch 42 due to minimal loss improvement.
</pre></div></div>
</div>
<p>Since we are using the filtered version of the data, if you want to obtain a full set of genes for the final dataset, you can use the following code:</p>
<p>```python</p>
<p>adata1 = sc.read_10x_mtx(data_path + ‘/square_002um/filtered_feature_bc_matrix’) adata1 = adata1[so.df[so.df.in_tissue == 1][‘barcode’]] weight_to_celltype = su.calculate_weight_to_celltype(adatas_final, adata1, cells_final, so) adata_sc_final2 = su.get_finaldata(adata1, adatas_final, spot_cell_dic, weight_to_celltype, cells_before_ml, groups_combined, pct_toml_dic, nonzero_indices_dic, spots_X_dic = None, nonzero_indices_toml = nonzero_indices_toml, cells_before_ml_x = None, so = so)
adata_sc_final2</p>
<p>we create our own pixels-to-cell plot and map it to the original staining image! This should be amazing! Note that saving to a file might take a relatively long time. Use <code class="docutils literal notranslate"><span class="pre">save=None</span></code> or <code class="docutils literal notranslate"><span class="pre">save=False</span></code> if saving is not needed. High DPI settings may also increase processing time; the recommended DPI is 1500.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">su</span><span class="o">.</span><span class="n">make_pixels_cells</span><span class="p">(</span><span class="n">so</span><span class="p">,</span> <span class="n">adata</span><span class="p">,</span> <span class="n">cells_before_ml</span><span class="p">,</span> <span class="n">spot_cell_dic</span><span class="p">,</span> <span class="n">spots_id_dic</span><span class="p">,</span> <span class="n">spots_id_dic_prop</span><span class="p">,</span> <span class="n">nonzero_indices_dic</span><span class="p">)</span>
<span class="n">su</span><span class="o">.</span><span class="n">plot_results</span><span class="p">(</span><span class="n">so</span><span class="o">.</span><span class="n">image_temp</span><span class="p">(),</span> <span class="n">so</span><span class="o">.</span><span class="n">pixels_cells</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">1500</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="n">save_path</span> <span class="o">+</span> <span class="s1">&#39;final_results.pdf&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 23340/23340 [26:17&lt;00:00, 14.80it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_40_1.png" src="../../_images/tutorials_notebooks_Tutorial_Mousebrian_full_40_1.png" />
</div>
</div>
<p>Finally, we create our own pixels-to-cell plot and map it to the original staining image! This should be amazing! Note that saving to a file might take a relatively long time. Use <code class="docutils literal notranslate"><span class="pre">save=None</span></code> or <code class="docutils literal notranslate"><span class="pre">save=False</span></code> if saving is not needed. High DPI settings may also increase processing time; the recommended DPI is 1500.</p>
<p>Here, <code class="docutils literal notranslate"><span class="pre">adata_sc_final.obs</span></code> contains five columns:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">cell_cluster</span></code>: Denotes the cluster assignment from the iteration.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cos_simularity</span></code>: The cosine similarity of the cell’s gene expression with the average expression of its cluster.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cell_size</span></code>: The number of <code class="docutils literal notranslate"><span class="pre">2um</span></code> spots occupied by this cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">array_row</span></code> and <code class="docutils literal notranslate"><span class="pre">array_col</span></code>: The absolute coordinates of the cell, matching the spot locations provided by 10x.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc_final</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">get_finaldata</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">adatas_final</span><span class="p">,</span> <span class="n">spot_cell_dic</span><span class="p">,</span> <span class="n">weights_record</span><span class="p">,</span> <span class="n">cells_before_ml</span><span class="p">,</span> <span class="n">groups_combined</span><span class="p">,</span>
                               <span class="n">pct_toml_dic</span><span class="p">,</span>  <span class="n">nonzero_indices_dic</span><span class="p">,</span> <span class="n">spots_X_dic</span><span class="p">,</span> <span class="n">cells_before_ml_x</span> <span class="o">=</span> <span class="n">cells_before_ml_x</span><span class="p">,</span> <span class="n">so</span> <span class="o">=</span> <span class="n">so</span><span class="p">)</span>
<span class="n">adata_sc_final</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 56858 × 9970
    obs: &#39;cell_cluster&#39;, &#39;cos_simularity&#39;, &#39;cell_size&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_counts&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;
    uns: &#39;spatial&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;so.pkl&quot;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">so</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>

<span class="n">adata_sc_final</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;adata_sc_final.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata1</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_mtx</span><span class="p">(</span><span class="n">data_path</span> <span class="o">+</span><span class="s1">&#39;binned_outputs/square_002um/filtered_feature_bc_matrix&#39;</span><span class="p">)</span>
<span class="n">adata1</span> <span class="o">=</span> <span class="n">adata1</span><span class="p">[</span><span class="n">so</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="n">so</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">in_tissue</span> <span class="o">==</span> <span class="mi">1</span><span class="p">][</span><span class="s1">&#39;barcode&#39;</span><span class="p">]]</span>
<span class="n">weight_to_celltype</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">calculate_weight_to_celltype</span><span class="p">(</span><span class="n">adatas_final</span><span class="p">,</span> <span class="n">adata1</span><span class="p">,</span> <span class="n">cells_final</span><span class="p">,</span> <span class="n">so</span><span class="p">)</span>
<span class="n">adata_sc_final2</span> <span class="o">=</span> <span class="n">su</span><span class="o">.</span><span class="n">get_finaldata</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adatas_final</span><span class="p">,</span> <span class="n">spot_cell_dic</span><span class="p">,</span> <span class="n">weight_to_celltype</span><span class="p">,</span> <span class="n">cells_before_ml</span><span class="p">,</span>
                                   <span class="n">groups_combined</span><span class="p">,</span> <span class="n">pct_toml_dic</span><span class="p">,</span> <span class="n">nonzero_indices_dic</span><span class="p">,</span> <span class="n">spots_X_dic</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                                   <span class="n">nonzero_indices_toml</span> <span class="o">=</span> <span class="n">nonzero_indices_toml</span><span class="p">,</span> <span class="n">cells_before_ml_x</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">so</span> <span class="o">=</span> <span class="n">so</span><span class="p">)</span>
<span class="n">adata_sc_final2</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 56867/56867 [01:27&lt;00:00, 649.02it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 56859 × 19059
    obs: &#39;cell_cluster&#39;, &#39;cos_simularity&#39;, &#39;cell_size&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_counts&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;
    uns: &#39;spatial&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc_final2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">save_path</span> <span class="o">+</span> <span class="s2">&quot;adata_sc_final2.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="unroll_example.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Unrolling Tutorial</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="Tutorial_Mousebrian.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Soft Segmentation Tutorial (Lite version)</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; Mitra Lab, Washington University in St. Louis
            </div>
            Made with 
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Soft Segmentation Tutorial (Full version)</a><ul>
<li><a class="reference internal" href="#Dataset-Information">Dataset Information</a></li>
<li><a class="reference internal" href="#Important-Downloads">Important Downloads</a><ul>
<li><a class="reference internal" href="#Start-Here">Start Here</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../../_static/jquery.js?v=5d32c60e"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="../../_static/documentation_options.js?v=f5cff4aa"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/furo.js?v=46bd48cc"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/js/hoverxref.js"></script>
    <script src="../../_static/js/tooltipster.bundle.min.js"></script>
    <script src="../../_static/js/micromodal.min.js"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>